test/e2e/e2e_suite_test.go

```diff
@@ -120,60 +122,36 @@ var _ = SynchronizedBeforeSuite(func() []byte {
 		// Install Ironic
 		By("Installing Ironic")
 		kustomization := e2eConfig.GetVariable("IRONIC_KUSTOMIZATION")
-		manifest, err := buildKustomizeManifest(kustomization)
-		Expect(err).NotTo(HaveOccurred())
-		err = clusterProxy.Apply(ctx, manifest)
-		Expect(err).NotTo(HaveOccurred())
-
-		ironicDeployment := &v1.Deployment{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      "ironic",
-				Namespace: "baremetal-operator-system",
-			},
-		}
-		// Wait for it to become available
-		framework.WaitForDeploymentsAvailable(ctx, framework.WaitForDeploymentsAvailableInput{
-			Getter:     clusterProxy.GetClient(),
-			Deployment: ironicDeployment,
-		}, e2eConfig.GetIntervals("ironic", "wait-deployment")...)
-		// Set up log watcher
-		framework.WatchDeploymentLogsByName(ctx, framework.WatchDeploymentLogsByNameInput{
-			GetLister:  clusterProxy.GetClient(),
-			Cache:      clusterProxy.GetCache(ctx),
-			ClientSet:  clusterProxy.GetClientSet(),
-			Deployment: ironicDeployment,
-			LogPath:    filepath.Join(artifactFolder, "logs", ironicDeployment.GetNamespace()),
+		err := InstallIronicBMO(ctx, &InstallIronicBMOInput{
```

Comment:
If you want to refactor this (and I suggest not to do it in this PR), please make the function more generic and get rid of the env file input. Just apply the kustomization as it is.

Reply:
The env file is needed, because we need different ips for different clusters. BMO (and later on, ironic) can be upgraded from one version to another, but the ip should stay the same. Otoh, the same BMO version can be installed in both clusters, but the ips need to be different.
Not having the env file configurable will require us to have different kustomizations for all of these cases, which I find too much for just simple changes.

Reply:
IMO different kustomizations is much, much better. By reusing this one file you save a few lines, but at the same time you increase the complexity and basically require developers to read the code *instead* of the configuration.

- How do you know what is used before upgrade?
  - You find the kustomization that is used. It does not have any env file...
  - You check the code for how it is installed.
  - You find that it copies the file from some other place.
  - You find the actual configuration used.
- How do you know what is used after upgrade? Repeat the steps.
- How do you know what is used in non-upgrade tests? Repeat the steps.

See how much effort goes into finding out these very simple things?

Compare now with "wasting" maybe 50 lines by duplicating the file:

- How do you know what is used before upgrade?
  - You find the kustomization that is used. There is your answer.
- Same process for after upgrade and non-upgrade tests.

At the same time, this allows developers and users to use it in a flexible way. Maybe we want to try using a completely different config before and after upgrade (different IP)? Maybe we want the exact same config (basically no-op "upgrade")? We cannot know all the different ways this will be used which is why flexibility is so important.

To summarize: I find all this complex behavior and code "too much for just simple changes" :wink: 
Static config is not dangerous. DRY is not always the answer. I would say KISS is much more important.

Reply:
Just want to make a comment here that while I agree with everything you mentioned, it's worth to note that if we decide to duplicate the kustomization, we don't simply do it once. In this case, we would need two new kustomizations for BMO in the upgrade cluster, and one for ironic (+one more when we have ironic upgrade test). Imo that's not simple enough that we can say KISS was kept. 
Anyway, fortunately since we're going with separating the job to two, this dilemma does not need to happen anymore :smile:

Reply:
It is literally just a config file with <10 lines. That is KISS if anything. If the config is the same before and after upgrade, then you can even make a common overlay/component that adds the config and you would get just a kustomization.yaml that "assembles" the config with the rest of the manifests.

It is not an issue any more for CI but I'm trying to explain that this is exactly how I would do it in a real situation. Let's say I want to run this in a homelab. I definitely do not add code for copying and changing files back and forth to get the minimum possible line count. What I would do is to create a separate config for each environment. Nice and clean.